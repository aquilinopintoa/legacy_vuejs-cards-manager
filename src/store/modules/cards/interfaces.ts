export interface CardRawInterface {
    title: string;
    description: string;
    url: string;

}

export interface CardInterface {
    id: string;
    title: string;
    description: string;
    url: string;
    createdAt: string;
}
