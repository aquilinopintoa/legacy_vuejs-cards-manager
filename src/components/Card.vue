<template>
  <v-container>
    <v-layout
      text-xs-center
    >
        <v-card>
            <v-img
                src="https://cdn.vuetifyjs.com/images/cards/desert.jpg"
                aspect-ratio="2.75"
                ></v-img>

            <v-card-title primary-title>
                <div>
                    <h3 class="headline mb-0">Kangaroo Valley Safari</h3>
                    <div> {{ card_text }} </div>
                </div>
            </v-card-title>

            <v-card-actions>
                <v-btn flat color="orange">Share</v-btn>
                <v-btn flat color="orange">Explore</v-btn>
            </v-card-actions>
        </v-card>
    </v-layout>
  </v-container>
</template>

<script>
export default {
    data: () => ({
    })
};
</script>

<style>

</style>
