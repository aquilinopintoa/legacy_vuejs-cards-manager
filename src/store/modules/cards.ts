import moment from 'moment';
import Vue from 'vue';

const DEFAULT_URL = 'https://cdn.vuetifyjs.com/images/cards/desert.jpg';

export interface CardRawInterface {
    title: string;
    description: string;
    url: string;

}

export interface CardInterface {
    id: string;
    title: string;
    description: string;
    url: string;
    created_at: string;

}

export default {
    namespaced: true,
    state: {
        cards: {}
    },

    getters: {
        get: (state: any) =>
            (id: string): CardInterface | CardInterface[] => {
                if (id) {
                    return state.cards[id];
                } else {
                    return Object.values(state.cards);
                }
            }
    },

    mutations: {
        set (store: any, card: CardInterface) {
            Vue.set(store.cards, card.id, card);
        },
        reset (store: any) {
            store.cards = {};
        }
    },

    actions: {
        async create (context: any, cardRaw: CardRawInterface) {
            const createdAt = moment().format();
            const card = {
                id: `${cardRaw.title}${createdAt}`,
                created_at: createdAt,
                ...cardRaw,
                url: cardRaw.url || DEFAULT_URL
            };

            await context.commit('set', card);
        },
        async remove (context: any, id: string) {
            const previous = context.getters.get;
            const removedTarget = previous.filter((card: CardInterface) => card.id !== id);

            await context.commit('reset');

            await Promise.all(removedTarget.map((card: CardInterface) => context.commit('set', card)));
        }
    }
};
