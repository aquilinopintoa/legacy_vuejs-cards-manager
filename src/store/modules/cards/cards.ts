import getters from './state/getters';
import mutations from './state/mutations';
import actions from './state/actions';

export default {
    namespaced: true,

    state: {
        cards: {}
    },

    getters,
    mutations,
    actions
};
