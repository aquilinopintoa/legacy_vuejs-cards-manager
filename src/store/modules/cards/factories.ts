import moment from 'moment';
import { CardRawInterface, CardInterface} from './interfaces';

const DEFAULT_URL = 'https://cdn.vuetifyjs.com/images/cards/desert.jpg';

export const FactoryCard = (raw: CardRawInterface): CardInterface => {
    const createdAt = moment().format();
    return {
        id: `${raw.title}${createdAt}`,
        createdAt,
        ...raw,
        url: raw.url || DEFAULT_URL
    };
};
