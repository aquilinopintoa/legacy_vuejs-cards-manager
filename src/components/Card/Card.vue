<template>
  <v-container>
    <v-layout
      text-xs-center
    >
        <v-card 
            width="100%" 
            @mouseover="handlerHover"
            @mouseout="handlerHover">
            <v-img
                :src="card.url"
                aspect-ratio="2.75"
                class="align-end"
                >
                <v-container
                    fluid
                    pa-2
                    align-end
                    justify-start
                  >
                    <div class="headline white--text text-xs-left">{{ card.title }}</div>
                  </v-container>
            </v-img>
            <v-card-title primary-title>
                <div>
                    <div> {{ card.description }} </div>
                </div>
            </v-card-title>

            <v-card-actions class="justify-end">
                <div v-show="hover">
                    <v-btn 
                        v-for="(action, index) of actions" 
                        :key="index" 
                        flat
                        @click="action.onAction(card)" 
                        :color="action.labelColor">{{action.label}}</v-btn>
                </div>
            </v-card-actions>
        </v-card>
    </v-layout>
  </v-container>
</template>

<script lang="ts">
import { CardInterface } from '@/store/modules/cards/cards';
import { Component, Vue, Prop } from 'vue-property-decorator';

export interface CardActionInterface {
    label: string;
    labelColor: string;
    onAction: (card: CardInterface) => void;
}

@Component({
})
export default class Card extends Vue {
    @Prop({required: true})
    private readonly card!: CardInterface;
    @Prop({default: []})
    private readonly actions!: CardActionInterface[];
    private hover: boolean = false;

    private handlerHover (): void {
        this.hover = !this.hover;
    }
}
</script>

<style>
.min-height {
    min-height: 52px
}
</style>
