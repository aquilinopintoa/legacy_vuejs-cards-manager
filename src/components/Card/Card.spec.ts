import { shallowMount } from '@vue/test-utils';
import Vue from 'vue';
import Vuetify from 'vuetify';
import Card from '@/components/Card/Card.vue';
import { CardInterface } from '@/store/modules/cards/interfaces';
import { FactoryCard } from '@/store/modules/cards/factories';

Vue.use(Vuetify);

describe('Card.vue', () => {
    let wrapper: any = null;
    const card: CardInterface = FactoryCard({
        title: 'test',
        description: 'test-description',
        url: 'https://test.test.co'
    });

    beforeEach(() => {
        wrapper = shallowMount(Card, {
            propsData: { card,  actions: []},
        });
    });

    it('returns an error for a valid code', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('renders props.msg when passed', () => {
        expect(wrapper.text()).toMatch(card.description);
    });
});
