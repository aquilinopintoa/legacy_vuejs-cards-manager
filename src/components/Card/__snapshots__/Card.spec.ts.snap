// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card.vue returns an error for a valid code 1`] = `
<v-container-stub tag="div">
  <v-layout-stub tag="div" text-xs-center="">
    <v-card-stub tag="div" width="100%">
      <v-img-stub aspectratio="2.75" src="https://test.test.co" position="center center" transition="fade-transition" class="align-end">
        <v-container-stub tag="div" fluid="" pa-2="" align-end="" justify-start="">
          <div class="headline white--text text-xs-left">test</div>
        </v-container-stub>
      </v-img-stub>
      <v-card-title-stub primarytitle="true">
        <div>
          <div> test-description </div>
        </div>
      </v-card-title-stub>
      <v-card-actions-stub class="justify-end">
        <div style="display: none;"></div>
      </v-card-actions-stub>
    </v-card-stub>
  </v-layout-stub>
</v-container-stub>
`;
